File: shared_resources/src/main/java/com/thewizrd/shared_resources/helpers/MultiChoiceAction.java
package com.thewizrd.shared_resources.helpers;

public final class MultiChoiceAction extends Action {
    private int value;

    private MultiChoiceAction() {
        super();
    }

    public MultiChoiceAction(Actions action) {
        super(action);
    }

    public MultiChoiceAction(Actions action, int choice) {
        super(action);
        this.value = choice;
    }

    public int getChoice() {
        return value;
    }

    public void setChoice(int value) {
        int maxStates = getNumberOfStates();
        int choice = value % maxStates;
        if (choice < 0) choice += maxStates;

        this.value = choice;
    }

    public int getNumberOfStates() {
        switch (getAction()) {
            case WIFI:
            case BLUETOOTH:
            case MOBILEDATA:
            case LOCATION:
            case TORCH:
            case LOCKSCREEN:
            case VOLUME:
            default:
                return 1;
            case DONOTDISTURB:
                return DNDChoice.values().length;
            case RINGER:
                return RingerChoice.values().length;
        }
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/helpers/MultiChoiceAction.kt
package com.thewizrd.shared_resources.helpers

class MultiChoiceAction : Action {
    private var value = 0

    constructor(action: Actions) : super(action)
    constructor(action: Actions, choice: Int) : super(action) {
        this.choice = choice
    }

    var choice: Int
        get() = value
        set(value) {
            val maxStates = numberOfStates
            var choice = value % maxStates
            if (choice < 0) choice += maxStates

            this.value = choice
        }

    val numberOfStates: Int
        get() {
            return when (actionType) {
                Actions.WIFI,
                Actions.BLUETOOTH,
                Actions.MOBILEDATA,
                Actions.LOCATION,
                Actions.TORCH,
                Actions.LOCKSCREEN,
                Actions.VOLUME -> 1
                Actions.DONOTDISTURB -> DNDChoice.values().size
                Actions.RINGER -> RingerChoice.values().size
                else -> 1
            }
        }
}